internal/session/manager.go

```diff
@@ -0,0 +1,355 @@
+	if err != nil {
+		return sessionManager, err
+	}
+
+	sessionManager.localSessionID = (model.SessionID)(randomBytes[:8])
+
+	localKey, err := NewKeySource()
+	if err != nil {
+		return sessionManager, err
+	}
+
+	k, err := sessionManager.ActiveKey()
+	if err != nil {
+		return sessionManager, err
+	}
+	k.local = localKey
+	return sessionManager, nil
+}
+
+// LocalSessionID gets the local session ID as bytes.
+func (m *Manager) LocalSessionID() []byte {
+	defer m.mu.Unlock()
+	m.mu.Lock()
+	return m.localSessionID[:]
+}
+
+// RemoteSessionID gets the remote session ID as bytes.
+func (m *Manager) RemoteSessionID() []byte {
+	defer m.mu.Unlock()
+	m.mu.Lock()
+	rs := m.remoteSessionID
+	if !rs.IsNone() {
+		val := rs.Unwrap()
+		return val[:]
+	}
+	return nil
+}
+
+// IsRemoteSessionIDSet returns whether we've set the remote session ID.
+func (m *Manager) IsRemoteSessionIDSet() bool {
+	defer m.mu.Unlock()
+	m.mu.Lock()
+	return !m.remoteSessionID.IsNone()
+}
+
+// ErrNoRemoteSessionID indicates we are missing the remote session ID.
+var ErrNoRemoteSessionID = errors.New("missing remote session ID")
+
+// NewACKForPacket creates a new ACK for the given packet.
+func (m *Manager) NewACKForPacket(packet *model.Packet) (*model.Packet, error) {
+	defer m.mu.Unlock()
+	m.mu.Lock()
+	if m.remoteSessionID.IsNone() {
+		return nil, ErrNoRemoteSessionID
+	}
+	p := &model.Packet{
+		Opcode:          model.P_ACK_V1,
+		KeyID:           m.keyID,
+		PeerID:          [3]byte{},
+		LocalSessionID:  m.localSessionID,
+		ACKs:            []model.PacketID{packet.ID},
+		RemoteSessionID: m.remoteSessionID.Unwrap(),
+		ID:              0,
+		Payload:         []byte{},
+	}
+	return p, nil
+}
+
+// NewPacket creates a new packet for this session.
+func (m *Manager) NewPacket(opcode model.Opcode, payload []byte) (*model.Packet, error) {
+	defer m.mu.Unlock()
+	m.mu.Lock()
+	// TODO: consider unifying with ACKing code
+	packet := model.NewPacket(
+		opcode,
+		m.keyID,
+		payload,
+	)
+	copy(packet.LocalSessionID[:], m.localSessionID[:])
+	pid, err := func() (model.PacketID, error) {
+		if opcode.IsControl() {
+			return m.localControlPacketIDLocked()
+		} else {
+			return m.localDataPacketIDLocked()
+		}
+	}()
+	if err != nil {
+		return nil, err
+	}
+	packet.ID = pid
+	if !m.remoteSessionID.IsNone() {
+		packet.RemoteSessionID = m.remoteSessionID.Unwrap()
+	}
+	return packet, nil
+}
+
+var ErrExpiredKey = errors.New("expired key")
+
+// LocalDataPacketID returns an unique Packet ID for the Data Channel. It
+// increments the counter for the local data packet ID.
+func (m *Manager) LocalDataPacketID() (model.PacketID, error) {
+	defer m.mu.Unlock()
+	m.mu.Lock()
+	return m.localDataPacketIDLocked()
+}
+
+// localDataPacketIDLocked returns an unique Packet ID for the Data Channel. It
+// increments the counter for the local data packet ID.
+func (m *Manager) localDataPacketIDLocked() (model.PacketID, error) {
+	pid := m.localDataPacketID
+	if pid == math.MaxUint32 {
```

Comment:
```suggestion
	if pid >= math.MaxUint32 {
```

Reply:
I think this is not correct right? PacketID is a uint32 so it should not overflow.
This is a very improbable error, but the only way of returning it meaningfully is to error on the last usable PacketID, I think.

Reply:
I tend to like expressing this kind of checks in a more defensive way which makes them more robust to refactoring like "let's change this to int64 now", but feel free to ignore if you don't like it.